### PR TITLE
Lean4 solution by badly-drawn-wizards

### PR DESCRIPTION
**Pull request by badly-drawn-wizards**
## Description
An implementation in the Lean4 proof assistant.

## Contributing requirements
* [x] I read the contribution guidelines in CONTRIBUTING.md.
* [x] I placed my solution in the correct solution folder.
* [x] I added a README.md with the right badge(s).
* [x] I added a Dockerfile that builds and runs my solution.
* [x] I selected `drag-race` as the target branch.
* [x] All code herein is licensed compatible with BSD-3.

### CODE REVIEW

File: PrimeLean4/solution_1/PrimeLean4.lean [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  let start_ms := ← IO.monoMsNow
+  let stop_ms := start_ms + dur_ms
+  let mut iterations : Nat := 0
+  let mut last_ms := start_ms
+  for _ in Loop.mk do
+    last_ms := (← IO.monoMsNow)
+    if stop_ms ≤ last_ms then break
+    don't_optimize $ impl.run size
+    iterations := iterations + 1
+  pure ⟨impl.toImplMetadata, iterations, last_ms - start_ms⟩
```
Comment (rbergen) [inline PrimeLean4/solution_1/PrimeLean4.lean:101]: For an implementation to be marked as faithful, it needs to recreate the sieve class (or equivalent) instance from scratch for every run. This implementation seems to operate on a single `Impl` instance that is reused within the timed loop. Can you please take me through how this implementation handles the rule I just referred to?

File: PrimeLean4/solution_1/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+# Install lean toolchain early so that changes in source code don't
+# trigger a redownload of lean
+RUN elan toolchain install leanprover/lean4:4.0.0-m2
+RUN apt-get install build-essential libgmp-dev -y
+
+WORKDIR /app
+COPY . .
+
+RUN leanpkg configure
+RUN leanpkg build bin
```
Comment (rbergen) [inline PrimeLean4/solution_1/Dockerfile]: Please consolidate the various series of `RUN` directives to one directive per set of commands.

File: PrimeLean4/solution_1/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ENTRYPOINT [ ./build/bin/PrimeLean4 ]
```
Comment (rbergen) [inline PrimeLean4/solution_1/Dockerfile]: When I `docker run` the image, nothing happens:
```
$ docker run -ti --rm lean4
$
```

File: PrimeLean4/solution_1/PrimeLean4.lean (hunk 1 above)
Comment (badly-drawn-wizards) [reply PrimeLean4/solution_1/PrimeLean4.lean:101]: Hi. This is a confusion of naming between FP and OOP communities. Classes here are typeclasses, which are distinct from the notion of classes in OOP. The equivalent of the sieve class being recreated here is the recreation of the array of primes each run. If this is not deemed equivalent, then I don't mind unmarking it as unfaithful.

Comment (badly-drawn-wizards) [reply PrimeLean4/solution_1/PrimeLean4.lean:101]: You can think of the classes here as closer to Rust's traits.

File: PrimeLean4/solution_1/Dockerfile (hunk 3 above)
Comment (badly-drawn-wizards) [reply PrimeLean4/solution_1/Dockerfile]: Ah, forgot quotes. Is fixed now.

File: PrimeLean4/solution_1/PrimeLean4.lean (hunk 1 above)
Comment (rbergen) [reply PrimeLean4/solution_1/PrimeLean4.lean:101]: Ok, thanks for clarifying that.

File: PrimeLean4/solution_1/PrimeLean4.lean [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  run : Nat → array
+
+structure Output :=
+  impl : ImplMetadata
+  iterations : Nat
+  total_ms : Nat
+
+instance : ToString Output where
+  toString o := 
+    let tags := ";".intercalate (o.impl.tags.map toString)
```
Comment (rbergen) [inline PrimeLean4/solution_1/PrimeLean4.lean]: The separator that has to be used between tags is the comma (,), not the semicolon (;). The semicolon _is_ the correct separator for the fields in the output line (line 89).

**Review by rbergen — APPROVED** [review]
LGTM

**Review by marghidanu — APPROVED** [review]

**Pull request merged**